off_chain_data/application-go/listen.go

```diff
@@ -0,0 +1,93 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"offChainData/parser"
+	"offChainData/processor"
+	"offChainData/store"
+	"offChainData/utils"
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+
+	"github.com/hyperledger/fabric-gateway/pkg/client"
+	"google.golang.org/grpc"
+)
+
+func listen(clientConnection *grpc.ClientConn) {
+	id, options := newConnectOptions(clientConnection)
+	gateway, err := client.Connect(id, options...)
+	if err != nil {
+		panic(err)
+	}
+	defer func() {
+		gateway.Close()
+		fmt.Println("Gateway closed.")
+	}()
+
+	checkpointFile := utils.EnvOrDefault("CHECKPOINT_FILE", "checkpoint.json")
+	checkpointer, err := client.NewFileCheckpointer(checkpointFile)
+	if err != nil {
+		panic(err)
+	}
+	defer func() {
+		checkpointer.Close()
+		fmt.Println("Checkpointer closed.")
+	}()
+
+	fmt.Println("Start event listening from block", checkpointer.BlockNumber())
+	fmt.Println("Last processed transaction ID within block:", checkpointer.TransactionID())
+	if store.SimulatedFailureCount > 0 {
+		fmt.Printf("Simulating a write failure every %d transactions\n", store.SimulatedFailureCount)
+	}
+
+	ctx, close := signal.NotifyContext(context.Background(), os.Interrupt, syscall.SIGTERM)
+	defer func() {
+		close()
+		fmt.Println("Context closed.")
+	}()
+
+	network := gateway.GetNetwork(channelName)
+	blocks, err := network.BlockEvents(
+		ctx,
+		// Used only if there is no checkpoint block number.
+		// Order matters. WithStartBlock must be set before
+		// WithCheckpoint to work.
+		client.WithStartBlock(0),
+		client.WithCheckpoint(checkpointer),
+	)
+	if err != nil {
+		panic(err)
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	go func() {
+		defer wg.Done()
+
+		for blockProto := range blocks {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				blockProcessor := processor.NewBlock(
+					parser.ParseBlock(blockProto),
+					checkpointer,
+					store.ApplyWritesToOffChainStore,
+					channelName,
+				)
+
+				if err := blockProcessor.Process(); err != nil {
+					fmt.Println("\033[31m[ERROR]\033[0m", err)
+					return
+				}
+			}
+		}
+	}()
+
+	wg.Wait()
```

Comment:
I'm not sure of the reason for using a wait group for a single goroutine that you then wait on before continuing. Maybe this can just be inline code?

It should not be necessary to check the context while reading. If the context used for the `BlockEvents` call is closed, the returned channel should also get closed, which would mean the range loop will exit.

```suggestion
	for blockProto := range blocks {
		blockProcessor := processor.NewBlock(
			parser.ParseBlock(blockProto),
			checkpointer,
			store.ApplyWritesToOffChainStore,
			channelName,
		)

		if err := blockProcessor.Process(); err != nil {
			fmt.Println("\033[31m[ERROR]\033[0m", err)
			return
		}
	}

```

Reply:
You're right. It's not needed. There must've been a reason why I put it there while developing but I can't recall.

I removed it.